rangefeedbuffer: introduce a rangefeed buffer

irfansharif opened the pull request:
Buffer provides a thin memory-bounded buffer to sit on top of a rangefeed. It
accumulates raw rangefeed events[^1], which can be flushed out in timestamp
sorted order en-masse whenever the rangefeed frontier is bumped. If we
accumulate more events than the limit allows for, we error out to the caller.

We need such a thing in both #69614 and #69661.

[^1]: Rangefeed error events are propagated to the caller, checkpoint events
     are discarded.

Release note: None

First commit is from #71256. Co-authored-by: Arul Ajmani <arula@cockroachlabs.com>.

irfansharif committed:
rangefeedbuffer: introduce a rangefeed buffer

In both #69614 and #69661 we find ourselves reaching for a thin,
memory-bounded buffer to sit on top a rangefeed. We want to be able to
accumulate raw rangefeed events, and when the rangefeed frontier is
bumped, flush everything out en-masse in timestamp sorted order. Package
rangefeedbuffer introduces such a datastructure.

If we're past the configured memory limit before having observed a
frontier bump, we have little recourse -- since we receive checkpoints and
rangefeed values on the same stream, we're waiting for a future
checkpoint to flush out the accumulated memory. Given this, we simply
error out to the caller (expecting them to re-establish the rangefeed).

Release note: None

Co-authored-by: Arul Ajmani <arula@cockroachlabs.com>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/71225)
<!-- Reviewable:end -->

arulajmani reviewed (commented):
I haven't had a thorough look yet, but I don't think we can use checkpoint events here as they may not encompass the entire rangefeed span. I think this needs to work on frontier timestamps instead. 

Maybe we can change the buffer methods to: 
 ```
func (b *Buffer) Add(ctx context.Context, ev *roachpb.RangeFeedValue) error
func (b *Buffer) Advance(ctx context.Context, TS hlc.Timestamp)
```
We can then use the `OnFrontierAdvance` option from https://github.com/cockroachdb/cockroach/pull/71256/ to carve out a pattern like: 

```
r, err := f.RangeFeed(ctx, ...,
		func(ctx context.Context, value *roachpb.RangeFeedValue) {
			b.Add(ctx, value)
		},
		rangefeed.WithOnFrontierAdvance(func(ctx context.Context, ts hlc.Timestamp) {
		        b.Advance(ctx, ts)
		}),
```

Wdyt?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-:-MlQKOdZEpa7B5_tZbUJ:bjeolzf)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzK8vLCJTgNigi3aDj:-MlzK8vLCJTgNigi3aDk:bqmwfdl) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L69)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 	value := ev.Val.Value
> 	if !value.IsPresent() {
> 		value = ev.Val.PrevValue
> 	}
> ```
</details>

what is this case?
___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 119 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzJVcVEmnYiGd4Gn1L:-MlzJVcVEmnYiGd4Gn1M:b7j58e5) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L119)):*
> ```Go
> type entry struct {
> 	*roachpb.RangeFeedEvent
> 	hlc.Timestamp
> ```

Why separate timestamp when it's already a field in the event? Function seems better.

Data structure question: we only care about the ordering when we go to deal with a frontier. It feels like it would be simpler to just `sort.Sort` at Flush point and append during execution? That also allows us to end up freeing up the buffer after a burst and to avoid allocating a new one on flush. The flip side is if there are many flushes before a lot of data, you'll sort for each flush, but I really don't care about that case.  The bigger reason to do it is that I think it's simpler.  Consider:

```
type rangefeedEvents []*roachpb.RangefeedEvent

func (events rangefeedEvents) Len() int { return len(events) }
func (events rangefeedEvents) Swap(i, j int) { events[i], events[j] = events[j], events[i] }
func (events rangefeedEvents) Less(i, j int) bool { return eventTS(events[i].Timestamp).Less(eventTS(events[j])) }

func eventTS(ev *roachpb.RangefeedEvent) hlc.Timestamp { /* ... */ }

// Add adds the given rangefeed entry to the buffer.
func (b *Buffer) Add(ctx context.Context, ev *roachpb.RangeFeedEvent) error {
        // ...

	b.mu.Lock()
	defer b.mu.Unlock()

	if value.Timestamp.LessEq(b.mu.frontier) {
		// If the rangefeed entry is at a timestamp less than or equal to our
		// last known checkpoint, we don't need to record it.
		return nil
	}

	if len(b.mu.events) > b.limit {
		return ErrBufferLimitExceeded
	}
        b.mu.events = append(b.mu.events, ev)
	return nil
}


func (b *Buffer) Flush(
	ctx context.Context, frontier hlc.Timestamp,
) (events []*roachpb.RangeFeedEvent) {
	b.mu.Lock()
	defer b.mu.Unlock()

	if frontier.Less(b.mu.frontier) {
		log.Fatalf(ctx, "frontier timestamp regressed: saw %s, previously %s", frontier, b.mu.frontier)
	}
        sort.Sort(rangefeedEvents(b.mu.events))
        idx := sort.Search(len(b.mu.events), func(i int) bool {
            return !eventTS(b.mu.events).Less(frontier)
        })
        events = b.mu.events[:idx]
        b.mu.events = b.mu.events[idx:]
	b.mu.frontier = frontier
	return events
}
```


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzK8vLCJTgNigi3aDj:-MlzeUCS8_ZMUe3Yl1Ja:b29719q) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L69)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	value := ev.Val.Value
> 	if !value.IsPresent() {
> 		value = ev.Val.PrevValue
> 	}
> ```
what is this case?
</blockquote></details>

Just a way to extract a timestamp for a rangefeed value deletion event.
___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 119 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzJVcVEmnYiGd4Gn1L:-Mlztw1d3cO9P9MsG8n2:biai2fv) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L119)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Why separate timestamp when it's already a field in the event? Function seems better.

Data structure question: we only care about the ordering when we go to deal with a frontier. It feels like it would be simpler to just `sort.Sort` at Flush point and append during execution? That also allows us to end up freeing up the buffer after a burst and to avoid allocating a new one on flush. The flip side is if there are many flushes before a lot of data, you'll sort for each flush, but I really don't care about that case.  The bigger reason to do it is that I think it's simpler.  Consider:

```
type rangefeedEvents []*roachpb.RangefeedEvent

func (events rangefeedEvents) Len() int { return len(events) }
func (events rangefeedEvents) Swap(i, j int) { events[i], events[j] = events[j], events[i] }
func (events rangefeedEvents) Less(i, j int) bool { return eventTS(events[i].Timestamp).Less(eventTS(events[j])) }

func eventTS(ev *roachpb.RangefeedEvent) hlc.Timestamp { /* ... */ }

// Add adds the given rangefeed entry to the buffer.
func (b *Buffer) Add(ctx context.Context, ev *roachpb.RangeFeedEvent) error {
        // ...

	b.mu.Lock()
	defer b.mu.Unlock()

	if value.Timestamp.LessEq(b.mu.frontier) {
		// If the rangefeed entry is at a timestamp less than or equal to our
		// last known checkpoint, we don't need to record it.
		return nil
	}

	if len(b.mu.events) > b.limit {
		return ErrBufferLimitExceeded
	}
        b.mu.events = append(b.mu.events, ev)
	return nil
}


func (b *Buffer) Flush(
	ctx context.Context, frontier hlc.Timestamp,
) (events []*roachpb.RangeFeedEvent) {
	b.mu.Lock()
	defer b.mu.Unlock()

	if frontier.Less(b.mu.frontier) {
		log.Fatalf(ctx, "frontier timestamp regressed: saw %s, previously %s", frontier, b.mu.frontier)
	}
        sort.Sort(rangefeedEvents(b.mu.events))
        idx := sort.Search(len(b.mu.events), func(i int) bool {
            return !eventTS(b.mu.events).Less(frontier)
        })
        events = b.mu.events[:idx]
        b.mu.events = b.mu.events[idx:]
	b.mu.frontier = frontier
	return events
}
```
</blockquote></details>

Good idea, done for both. One think I'm not sure about is this bit:

```
 events = b.mu.events[:idx]
 b.mu.events = b.mu.events[idx:]
```

Are we creating an ever-growing slice here, discarding only the prefix when flushing out events lesseq than the frontier timestamp?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzK8vLCJTgNigi3aDj:-Mlzuis2CdEYjXls0HO7:b-58duav) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L69)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Just a way to extract a timestamp for a rangefeed value deletion event.
</blockquote></details>

:thinking: the delete should have a timestamp, no?
___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 119 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzJVcVEmnYiGd4Gn1L:-MlzumEc7fW13lzuDx5v:b9ohd5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L119)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Good idea, done for both. One think I'm not sure about is this bit:

```
 events = b.mu.events[:idx]
 b.mu.events = b.mu.events[idx:]
```

Are we creating an ever-growing slice here, discarding only the prefix when flushing out events lesseq than the frontier timestamp?
</blockquote></details>

When the slice re-allocs, it'll be a new backing array. 


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzK8vLCJTgNigi3aDj:-Mlzwn0P8Ev_o5aJlGCf:bgkwups) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L69)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

:thinking: the delete should have a timestamp, no?
</blockquote></details>

Yea, isn't it ev.Val.PrevValue.Timestamp? That's what this code block was doing.
___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 119 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzJVcVEmnYiGd4Gn1L:-Mlzwsj--yWPiXhJY0YD:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L119)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

When the slice re-allocs, it'll be a new backing array. 
</blockquote></details>

Ack.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzK8vLCJTgNigi3aDj:-Mm-39HC5bqMSLmUDWOu:b-ufs8qm) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L69)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Yea, isn't it ev.Val.PrevValue.Timestamp? That's what this code block was doing.
</blockquote></details>

I don't think so. PrevValue will be empty unless you explicitly request it. It will also have the timestamp of the value that *preceded* the deletion. 

Can you write a test that exercises the deletion path? 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 6 of 6 files at r2, 3 of 4 files at r3, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @irfansharif)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzK8vLCJTgNigi3aDj:-Mm-CGK20rrDjK3W6anj:b-wf2qjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L69)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't think so. PrevValue will be empty unless you explicitly request it. It will also have the timestamp of the value that *preceded* the deletion. 

Can you write a test that exercises the deletion path? 

</blockquote></details>

I'm also confused by this. The `Value` should have a timestamp even if it is a deletion tombstone.
___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 111 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-Mm-A_gM7L7Xg1Pypa8Q:-Mm-A_gM7L7Xg1Pypa8R:bmevkur) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4a1e9c9ce60dbf98b319f022b03771114bb4ad/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L111)):*
> ```Go
> }
> 
> func eventTS(ev *roachpb.RangeFeedEvent) hlc.Timestamp {
> ```

@arulajmani and I were talking yesterday about whether this structure could be used in the `SQLWatcher` as well in place of the current `eventBuffer` he introduced over there. That use would be a little different in that it is zipping together two rangefeeds, but it has a lot of the same requirements. We discussed whether this buffer should be made slightly more generic to accommodate other uses cases, essentially by replacing `*roachpb.RangeFeedEvent` with a `type Event interface { Timestamp() hlc.Timestamp }`. 

We wanted to get your thoughts on that.


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzK8vLCJTgNigi3aDj:-Mm-O_Xo1r0KoqKobm-S:b-g0os3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L69)):*
Hm, looks like I misunderstood the APIs here; wrote a test at the top-level rangefeed package spelling out what the two timestamps should look like when writing/overwriting/deleting keys.

> It will also have the timestamp of the value that preceded the deletion.

Looks like this isn't true? I get an empty timestamp for the value preceding the deletion.
___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 111 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-Mm-A_gM7L7Xg1Pypa8Q:-Mm-P1G13lZvJM05xFwh:b-adavsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4a1e9c9ce60dbf98b319f022b03771114bb4ad/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L111)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

@arulajmani and I were talking yesterday about whether this structure could be used in the `SQLWatcher` as well in place of the current `eventBuffer` he introduced over there. That use would be a little different in that it is zipping together two rangefeeds, but it has a lot of the same requirements. We discussed whether this buffer should be made slightly more generic to accommodate other uses cases, essentially by replacing `*roachpb.RangeFeedEvent` with a `type Event interface { Timestamp() hlc.Timestamp }`. 

We wanted to get your thoughts on that.
</blockquote></details>

I don't mind the idea of hiding away the even through an interface. I imagine then in the SQLWatcher the events we'd store would be the decoded descriptor IDs (decoded as per the right table's schema), which feels cleaner than juggling two instances of this buffer; made that change. 

There's little in this library that's now "rangefeed" specific, though I expect it to mostly be used in conjunction with it. Suggestions for where it should be placed or what it should be named?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzK8vLCJTgNigi3aDj:-Mm-cDY_4YOf6CMgWDfP:b-bxwqnd) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L69)):*
> PrevValue will be empty unless you explicitly request it.


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzK8vLCJTgNigi3aDj:-Mm-cU1C9sw3KzpKD4MG:bp0oopr) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L69)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> PrevValue will be empty unless you explicitly request it.
</blockquote></details>

In TestRangefeedValueTimestamps I'm specifying `rangefeed.WithDiff()`, I'm seeing the previous values just fine -- the timestamps however are empty.


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzK8vLCJTgNigi3aDj:-Mm-cwA61sVFBC9ICOHL:b-jsqqmy) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L69)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

In TestRangefeedValueTimestamps I'm specifying `rangefeed.WithDiff()`, I'm seeing the previous values just fine -- the timestamps however are empty.
</blockquote></details>

https://github.com/cockroachdb/cockroach/blob/04170417a4314b222f4edaa117977b6a84207e0d/pkg/kv/kvserver/rangefeed/processor.go#L632-L644

Doesn't look like we set a prev-value timestamp above, unlike during the initial scan:

https://github.com/cockroachdb/cockroach/blob/fc803ee96268df35bbc7dfb9ef6f9cae37f1feb5/pkg/kv/kvclient/rangefeed/rangefeed.go#L258-L278


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzK8vLCJTgNigi3aDj:-Mm-dCuE9FG8dRhS6E_Z:b-nkc3zn) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L69)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

https://github.com/cockroachdb/cockroach/blob/04170417a4314b222f4edaa117977b6a84207e0d/pkg/kv/kvserver/rangefeed/processor.go#L632-L644

Doesn't look like we set a prev-value timestamp above, unlike during the initial scan:

https://github.com/cockroachdb/cockroach/blob/fc803ee96268df35bbc7dfb9ef6f9cae37f1feb5/pkg/kv/kvclient/rangefeed/rangefeed.go#L258-L278
</blockquote></details>

Ah, guess we didn't expose the PrevValue timestamp. That was probably smart from an API contract perspective.

https://github.com/cockroachdb/cockroach/blob/efb1a95071db1da80259367b2ef60ddbb6477427/pkg/kv/kvserver/rangefeed/processor.go#L625-L627

More commentary on https://github.com/cockroachdb/cockroach/blob/efb1a95071db1da80259367b2ef60ddbb6477427/pkg/roachpb/api.proto#L2384-L2388 probably wouldn't hurt.  


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzK8vLCJTgNigi3aDj:-Mm-dqwS4erqpc6V-9GU:blyfusz) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L69)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Ah, guess we didn't expose the PrevValue timestamp. That was probably smart from an API contract perspective.

https://github.com/cockroachdb/cockroach/blob/efb1a95071db1da80259367b2ef60ddbb6477427/pkg/kv/kvserver/rangefeed/processor.go#L625-L627

More commentary on https://github.com/cockroachdb/cockroach/blob/efb1a95071db1da80259367b2ef60ddbb6477427/pkg/roachpb/api.proto#L2384-L2388 probably wouldn't hurt.  
</blockquote></details>

Probably also worth reconciling those behaviors. That certainly was not intentional. 


<!-- Sent from Reviewable.io -->

irfansharif committed:
rangefeed: omit timestamps on previous values

There was a discrepancy with whether or not the previous contained the
timestamp of when the previous value was recorded. With the initial
scan, we populated the timestamp of the initial scan itself. With
regular diffs, we omitted the timestamp. This made for confusing
semantics (well, confusing at least for this author).

Release note: None

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 111 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-Mm-A_gM7L7Xg1Pypa8Q:-Mm0M1JL9m4zAh_lfhoT:b-q0xcow) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4a1e9c9ce60dbf98b319f022b03771114bb4ad/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L111)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

I don't mind the idea of hiding away the even through an interface. I imagine then in the SQLWatcher the events we'd store would be the decoded descriptor IDs (decoded as per the right table's schema), which feels cleaner than juggling two instances of this buffer; made that change. 

There's little in this library that's now "rangefeed" specific, though I expect it to mostly be used in conjunction with it. Suggestions for where it should be placed or what it should be named?
</blockquote></details>

Another thing that has come up in offline discussions with both @irfansharif and @nvanbenschoten, and I'd like to get more thoughts on here, is how we feel about generalizing this even further so that this can track multiple frontier timestamps. Specifically for the SQLWatcher, this would allow us to push the "combined frontier timestamp" into the buffer (as opposed to tracking it in a layer above). We could extend this interface a bit to make this work. 
___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 75 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-Mm0PhGJ1UkoEQsxar-u:-Mm0PhGJ1UkoEQsxar-v:b-yl8cuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7a0e52cd3ec9db40eac4f10b152bcce4244b34f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L75)):*
> ```Go
> // recorded (expected to monotonically increase), and future events with
> // timestamps less than or equal to it are discarded.
> func (b *Buffer) Flush(ctx context.Context, frontier hlc.Timestamp) (events []Event) {
> ```

In the EventBuffer from the `SQLWatcher` PR we were de-duplicating events from the buffer when flushing. Is that something we want to do here as well? 


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
TFTRs so far! Do we think this is close to getting merged? I want to build on top of this for the TODOs I've left over on https://github.com/cockroachdb/cockroach/pull/69614.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-:-Mm0SPAXDNMLTuC-TGAH:bpurezu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 69 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MlzK8vLCJTgNigi3aDj:-Mm0Qdm08_87VYwERwn6:b-l4364t) ([raw file](https://github.com/cockroachdb/cockroach/blob/efc15c50eb2b9b4eeecfd164c273d5f15e8d8a8f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L69)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Probably also worth reconciling those behaviors. That certainly was not intentional. 
</blockquote></details>

Added a commit.
___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 111 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-Mm-A_gM7L7Xg1Pypa8Q:-Mm0QqpY3XxUFG87A6WR:bbphwi9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4a1e9c9ce60dbf98b319f022b03771114bb4ad/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L111)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Another thing that has come up in offline discussions with both @irfansharif and @nvanbenschoten, and I'd like to get more thoughts on here, is how we feel about generalizing this even further so that this can track multiple frontier timestamps. Specifically for the SQLWatcher, this would allow us to push the "combined frontier timestamp" into the buffer (as opposed to tracking it in a layer above). We could extend this interface a bit to make this work. 
</blockquote></details>

I'm 👎 on it, especially if you can simply add as buffer events as decoded IDs. There are no other instances (to my knowledge) where we're maintaining multiple (n=2) rangefeeds to keep track of multiple frontier timestamps, so that sounds overgeneralized. What do you imagine that API would look like? Would it give us much over something like the following:

```go
type ... struct {
   rangefeedbuffer.Buffer // accumulates descpb.IDs
   zonesFrontierTS, descriptorFrontierTS hlc.Timestamp
}
```
___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 75 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-Mm0PhGJ1UkoEQsxar-u:-Mm0QfPcDxQJsJKPPARK:b-ltozl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7a0e52cd3ec9db40eac4f10b152bcce4244b34f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L75)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

In the EventBuffer from the `SQLWatcher` PR we were de-duplicating events from the buffer when flushing. Is that something we want to do here as well? 
</blockquote></details>

I had a TODO here, but decided ultimately against it -- easy enough to do at the caller + difficult now that we're taking in an opaque Event interface.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 7 files at r6, 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-:-MmJnbZAExOzmf_7X9so:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @irfansharif)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 111 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-Mm-A_gM7L7Xg1Pypa8Q:-MmJnVJJ4-3TJ0IQVXuI:b-rntivx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4a1e9c9ce60dbf98b319f022b03771114bb4ad/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L111)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

I'm 👎 on it, especially if you can simply add as buffer events as decoded IDs. There are no other instances (to my knowledge) where we're maintaining multiple (n=2) rangefeeds to keep track of multiple frontier timestamps, so that sounds overgeneralized. What do you imagine that API would look like? Would it give us much over something like the following:

```go
type ... struct {
   rangefeedbuffer.Buffer // accumulates descpb.IDs
   zonesFrontierTS, descriptorFrontierTS hlc.Timestamp
}
```
</blockquote></details>

+1 to what Irfan said.
___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 85 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MmJn6O2CU0ACX-3UCnS:-MmJn6O2CU0ACX-3UCnT:bfy0au6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7a0e52cd3ec9db40eac4f10b152bcce4244b34f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L85)):*
> ```Go
> 	// Accumulate all events with timestamps <= the given timestamp in sorted
> 	// order.
> 	sort.Sort(b.mu.events)
> ```

I _think_ that this is going to heap allocate because you have the interface implemented on a value. If you implement the interface on `*events` and then pass `sort.Sort(&b.mu.events)`, you can avoid that.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
I’m good on this. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 6 of 7 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-:-Mm1588Y45dty4iKBjJ5:bn9m8la)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 111 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-Mm-A_gM7L7Xg1Pypa8Q:-Mm14sqUBffCxLasbnTY:bh4evs8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4a1e9c9ce60dbf98b319f022b03771114bb4ad/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L111)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

+1 to what Irfan said.
</blockquote></details>

Had written this earlier but forgot to push publish. Seems like we're in agreement:

Leave generalizing refactors for later, when they’re obvious or better understood?


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
❤️ Thanks y'all!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-:-MmJuVTn-9EIMm5YhLod:b7at3bq)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @arulajmani and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 85 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MmJn6O2CU0ACX-3UCnS:-MmJtzJQ1tp3OoFjw4FU:baubvc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7a0e52cd3ec9db40eac4f10b152bcce4244b34f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L85)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I _think_ that this is going to heap allocate because you have the interface implemented on a value. If you implement the interface on `*events` and then pass `sort.Sort(&b.mu.events)`, you can avoid that.
</blockquote></details>

TIL, you're right:

```
$ go build -gcflags '-m' github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed/rangefeedbuffer
...
pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go:86:16: b.mu.events escapes to heap
```

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 85 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-MmJn6O2CU0ACX-3UCnS:-MmJv35zBrxPBFNQD77O:bal1x4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7a0e52cd3ec9db40eac4f10b152bcce4244b34f/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L85)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

TIL, you're right:

```
$ go build -gcflags '-m' github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed/rangefeedbuffer
...
pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go:86:16: b.mu.events escapes to heap
```

Done.
</blockquote></details>

an equivalent choice which maybe is worse is to keep the slice for the struct field and cast like: `sort.Sort((*events)(&b.mu.events))`. Also, I don't think you need to define the methods on the pointer to defeat the allocation as the pointer inherits the method set of the non-pointer. 


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71225)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go, line 111 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71225#-Mm-A_gM7L7Xg1Pypa8Q:-MmJypvM5fUV-7UDJ9TS:b-x3q8sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4a1e9c9ce60dbf98b319f022b03771114bb4ad/pkg/kv/kvclient/rangefeed/rangefeedbuffer/buffer.go#L111)):*
> What do you imagine that API would look like? Would it give us much over something like the following ...

We could structure it such that tracking of the frontier timestamp tracking is pushed inside this buffer, instead of the caller having to do it at the layer above. What I had in mind was something like: 

```
type Event interface {
   Timestamp() hlc.Timestamp
   Type() int                              // key into the array that tracks frontier timestamps
   FrontierAdvance()  bool     // true if the event corresponds to the frontier being advanced.
}
```

This would allow the SQLWatcher to construct different event types based on where the event was generated at (system.descriptors or system.zones) and not have to individually track the frontiers of these rangefeeds. 

All that being said, looks like this isn't super popular, so I'm fine this being structured as is. 


<!-- Sent from Reviewable.io -->

irfansharif commented:
CI + bazel infra flake: https://github.com/cockroachdb/cockroach/issues/71679, retrying.

bors r+
